### ISSUE TITLE
Using Autonity in console mode: fails to find Autonity contract; NULL pointer dereference

### ISSUE THREAD

**Issue by flofrie** [clearmatics/autonity#658]
Trying to use `autonity` instead of `geth` for attaching a console to TestNet:

`docker run -ti --rm clearmatics/autonity attach http://51.89.148.237:8545`

... to a Node **before** it's permissioned on the TestNet, so it hasn't yet started mining.

Seems to work nicely, except:

* right when starting, it reports: `contract binding failed:  Error: method handler crashed`
* On the server (node) side, it gives a slew of error messages, starting with `RPC method tendermint_getContractABI crashed: runtime error: invalid memory address or nil pointer dereference`

Probably due to Autonity contract address not yet known at that point. Should in any case fail more gracefully on the Node side.

**Comment by piersy** [comment]
Hi @flofrie, could you provide the slew of error messages, it sounds like its a stack trace and it would be helpful in tracking down the issue. Could you also let us know what version of autonity you are using?

**Comment by flofrie** [comment]
I'm afraid I can't anymore; it only came up when the Node wasn't permissioned yet, and at the time, I hadn't quite understood the severity of the problem. I only mentioned it on Slack, and Youssef recommended I create this Issue.

I will ask Ben to reproduce. (@piersy , @yazzaoui , can you please give @ben-clearmatics access to this repo ... or the Clearmatics organization ... since it seems I can't tag him here)

**Comment by ben-clearmatics** [comment]
I've just given this a go with one of my GCP docker instances, got these results:

On the client side the Autontiy JS console attached just fine with no errors at all, its at block 0 and the date is wrong but thats expected as it's not part of a network.

On the server side I get the following message when I Autonity attach:

WARN [08-27|13:22:32.934] can't get the contract ABI               err="leveldb: not found"

Other than that nothing changes, the node continues `Broadcasting consensus sync-me`

This was on autonity v0.4.1

**Comment by piersy** [comment]
@ben-clearmatics Thanks ben.
@flofrie I guess we can leave this ticket open for now and come back to it if we encounter this issue again. Without a stack trace or the ability to reproduce its going to be difficult to find the source of the issue.

**Comment by flofrie** [comment]
This may be another cased of this, reported by @AdamDynamic : https://support.clearmatics.com/helpdesk/tickets/209

Adam provided this:
```
lets go!
Import the genesis
INFO [09-11|10:47:18.591] Default Validator smart contract set
INFO [09-11|10:47:18.595] Maximum peer count                       ETH=50 LES=0 total=50
INFO [09-11|10:47:18.595] Smartcard socket not found, disabling    err="stat /run/pcscd/pcscd.comm: no such file or directory"
INFO [09-11|10:47:18.596] Allocated cache and file handles         database=/root/.ethereum/autonity/chaindata cache=16.00MiB handles=16
INFO [09-11|10:47:18.603] Writing custom genesis block
INFO [09-11|10:47:18.603] starting BFT consensus                   validators="[0x37345A93A4d745F1b3Ffc716dCBb5152f123DD37 - 50000] [0x52ECC0Bbe88507756735406cd31A4CfB7bcF717d - 50000] [0x55d2214E0e817BE995f2Dd7d95163e1a6E439618 - 50000] [0x5f3de5088f8A4eC170D7aE6e2dAB1ef4b9b39C82 - 50000] [0x7bbE26148A3CB2ce7a0f4f36c336382a8afD7323 - 50000] [0xd64986fb19c249776A68590B53fDf2d7752dd0CE - 50000] "
INFO [09-11|10:47:18.604] Persisted trie from memory database      nodes=1 size=172.00B time=92.41Âµs gcnodes=0 gcsize=0.00B gctime=0s livenodes=1 livesize=0.00B
WARN [09-11|10:47:18.607] storing chain config                     hash=0x843a11973abd446376ce9bc27de227a66913ad07e72da657102c714444db937c genesis="{ChainID: 444900 Homestead: 0 DAO: <nil> DAOSupport: false EIP150: 0 EIP155: 0 EIP158: 0 Byzantium: 0 Constantinople: 0 Petersburg: 0 Istanbul: 0, Muir Glacier: <nil>, Engine: tendermint}"
INFO [09-11|10:47:18.609] Successfully wrote genesis state         database=chaindata                          hash=843a11â€¦db937c
INFO [09-11|10:47:18.609] Allocated cache and file handles         database=/root/.ethereum/autonity/lightchaindata cache=16.00MiB handles=16
INFO [09-11|10:47:18.612] Writing custom genesis block
INFO [09-11|10:47:18.613] starting BFT consensus                   validators="[0x37345A93A4d745F1b3Ffc716dCBb5152f123DD37 - 50000] [0x52ECC0Bbe88507756735406cd31A4CfB7bcF717d - 50000] [0x55d2214E0e817BE995f2Dd7d95163e1a6E439618 - 50000] [0x5f3de5088f8A4eC170D7aE6e2dAB1ef4b9b39C82 - 50000] [0x7bbE26148A3CB2ce7a0f4f36c336382a8afD7323 - 50000] [0xd64986fb19c249776A68590B53fDf2d7752dd0CE - 50000] "
INFO [09-11|10:47:18.614] Persisted trie from memory database      nodes=1 size=172.00B time=73.353Âµs gcnodes=0 gcsize=0.00B gctime=0s livenodes=1 livesize=0.00B
WARN [09-11|10:47:18.616] storing chain config                     hash=0x843a11973abd446376ce9bc27de227a66913ad07e72da657102c714444db937c genesis="{ChainID: 444900 Homestead: 0 DAO: <nil> DAOSupport: false EIP150: 0 EIP155: 0 EIP158: 0 Byzantium: 0 Constantinople: 0 Petersburg: 0 Istanbul: 0, Muir Glacier: <nil>, Engine: tendermint}"
INFO [09-11|10:47:18.616] Successfully wrote genesis state         database=lightchaindata                          hash=843a11â€¦db937c
Let's start the Node
INFO [09-11|10:47:18.732] Starting pprof server                    addr=http://0.0.0.0:6060/debug/pprof
INFO [09-11|10:47:18.732] Enabling metrics collection
WARN [09-11|10:47:18.732] Async resolve started
INFO [09-11|10:47:18.736] Maximum peer count                       ETH=50 LES=0 total=50
INFO [09-11|10:47:18.737] Smartcard socket not found, disabling    err="stat /run/pcscd/pcscd.comm: no such file or directory"
INFO [09-11|10:47:18.738] Starting peer-to-peer node               instance=Autonity/v0.4.0-0bf95967-20200804/linux-amd64/go1.13.14
INFO [09-11|10:47:18.738] Allocated trie memory caches             clean=256.00MiB dirty=256.00MiB
INFO [09-11|10:47:18.738] Allocated cache and file handles         database=/root/.ethereum/autonity/chaindata cache=512.00MiB handles=524288
INFO [09-11|10:47:18.760] Initialised chain configuration          config="{ChainID: 444900 Homestead: 0 DAO: <nil> DAOSupport: false EIP150: 0 EIP155: 0 EIP158: 0 Byzantium: 0 Constantinople: 0 Petersburg: 0 Istanbul: 0, Muir Glacier: <nil>, Engine: tendermint}"
WARN [09-11|10:47:18.761] new backend with public key              addr=0x4b0d1DB4e02f568058Bbd1746F69dc3C6a94Ed2A
INFO [09-11|10:47:18.762] Initialising Ethereum protocol           versions="[64 63]" network=444900 dbversion=<nil>
WARN [09-11|10:47:18.762] Upgrade blockchain database version      from=<nil> to=7
INFO [09-11|10:47:18.765] Loaded most recent local header          number=0 hash=843a11â€¦db937c td=1 age=51y5mo6d
INFO [09-11|10:47:18.765] Loaded most recent local full block      number=0 hash=843a11â€¦db937c td=1 age=51y5mo6d
INFO [09-11|10:47:18.766] Loaded most recent local fast block      number=0 hash=843a11â€¦db937c td=1 age=51y5mo6d
INFO [09-11|10:47:18.767] Regenerated local transaction journal    transactions=0 accounts=0
INFO [09-11|10:47:18.776] Private-network mode enabled.
INFO [09-11|10:47:18.781] New local node record                    seq=1 id=ef5ff1e381633786 ip=127.0.0.1 udp=0 tcp=30303
INFO [09-11|10:47:18.781] Started P2P networking                   self="enode://5b271008b522adfed14025a1e6508dfdca53770dc789f75e2160153ffa467b9b91e74b255aa3906e1416e3cd1ce29f17caf347d26977e205d0fc1a76a972d067@127.0.0.1:30303?discport=0"
INFO [09-11|10:47:18.782] IPC endpoint opened                      url=/root/.ethereum/autonity.ipc
INFO [09-11|10:47:18.783] HTTP endpoint opened                     url=http://0.0.0.0:8545          cors=* vhosts=*
INFO [09-11|10:47:18.784] Reading Whitelist                        list="[enode://bc6c36b82db8066df9b3f8da30240f322a15ae049c592bf663c8c30d6c7d110127e3f8fc60da5c646daf5a4ef8769800f2e917b60d7d5062e6c7781408b57516@35.177.8.113:30308 enode://4220adad8c1b9b06baba249be3c90c9c7ee37f7575e01a91841f21fa6c9b60d57efb85f416157d5b305e2f53829cfc1550b5dbec02543252461db2f8228216e6@3.9.98.39:30310 enode://e923776167b2e7efecc2698eaddc1d47f44785f8b898d19f71b70bc31693a148fca5ca4a970f04f3444637b86e1225efe5257adf8c365ce182ade434dd6ffdf2@35.246.21.247:30303 enode://f1660e51bff264307bf03447567b319e3449bcc283900759ff4e959900fdcc2c60de2a6f3375972d640558abceaaaeea6b00e438c050d4fa7a6933a97091563a@35.189.64.207:30303 enode://cdf4df912f56b34eb74046033866f449dd086a1cdc4dad24cabdf4e58c331e389999b9d323d71179beb903eb31ec57ffd19cd2304e404c0731f279ba9ffe5b6d@34.105.163.137:30303 enode://4038d7575d99ec6a4c4b437849a26c2187bc63a7a29dd4991143fce2f6cf4738459930d5f3bc50f507a7e5ee41e158499c3eaab8da3a2fb0d54fb7ef679403ed@35.179.46.181:30309]"
INFO [09-11|10:47:18.784] Connecting to newly authorized peer      enode=enode://bc6c36b82db8066df9b3f8da30240f322a15ae049c592bf663c8c30d6c7d110127e3f8fc60da5c646daf5a4ef8769800f2e917b60d7d5062e6c7781408b57516@35.177.8.113:30308
INFO [09-11|10:47:18.784] Connecting to newly authorized peer      enode=enode://4220adad8c1b9b06baba249be3c90c9c7ee37f7575e01a91841f21fa6c9b60d57efb85f416157d5b305e2f53829cfc1550b5dbec02543252461db2f8228216e6@3.9.98.39:30310
INFO [09-11|10:47:18.784] Connecting to newly authorized peer      enode=enode://e923776167b2e7efecc2698eaddc1d47f44785f8b898d19f71b70bc31693a148fca5ca4a970f04f3444637b86e1225efe5257adf8c365ce182ade434dd6ffdf2@35.246.21.247:30303
INFO [09-11|10:47:18.785] Connecting to newly authorized peer      enode=enode://f1660e51bff264307bf03447567b319e3449bcc283900759ff4e959900fdcc2c60de2a6f3375972d640558abceaaaeea6b00e438c050d4fa7a6933a97091563a@35.189.64.207:30303
INFO [09-11|10:47:18.785] Connecting to newly authorized peer      enode=enode://cdf4df912f56b34eb74046033866f449dd086a1cdc4dad24cabdf4e58c331e389999b9d323d71179beb903eb31ec57ffd19cd2304e404c0731f279ba9ffe5b6d@34.105.163.137:30303
INFO [09-11|10:47:18.785] Connecting to newly authorized peer      enode=enode://4038d7575d99ec6a4c4b437849a26c2187bc63a7a29dd4991143fce2f6cf4738459930d5f3bc50f507a7e5ee41e158499c3eaab8da3a2fb0d54fb7ef679403ed@35.179.46.181:30309
INFO [09-11|10:47:18.786] WebSocket endpoint opened                url=ws://127.0.0.1:8546
ERROR[09-11|10:51:02.288] RPC method tendermint_getContractABI crashed: runtime error: invalid memory address or nil pointer dereference
goroutine 578 [running]:
github.com/clearmatics/autonity/rpc.(*callback).call.func1(0xc000c4f6a0, 0x19, 0xc000521d50)
	/autonity/build/_workspace/src/github.com/clearmatics/autonity/rpc/service.go:200 +0xb5
panic(0xff3280, 0x1b7bbc0)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/clearmatics/autonity/core.(*BlockChain).GetAutonityContract(...)
	/autonity/build/_workspace/src/github.com/clearmatics/autonity/core/blockchain.go:2240
github.com/clearmatics/autonity/consensus/tendermint/backend.(*Backend).GetContractABI(0xc0002fc900, 0x8, 0xc000ca4140)
	/autonity/build/_workspace/src/github.com/clearmatics/autonity/consensus/tendermint/backend/backend.go:438 +0x26
github.com/clearmatics/autonity/consensus/tendermint/backend.(*API).GetContractABI(0xc0001cfde0, 0x0, 0x0)
	/autonity/build/_workspace/src/github.com/clearmatics/autonity/consensus/tendermint/backend/api.go:62 +0x37
reflect.Value.call(0xc000541440, 0xc000011cc8, 0x13, 0x115c570, 0x4, 0xc000c68c00, 0x1, 0x2, 0xc000521bc8, 0x46027c, ...)
	/usr/local/go/src/reflect/value.go:460 +0x5f6
reflect.Value.Call(0xc000541440, 0xc000011cc8, 0x13, 0xc000c68c00, 0x1, 0x2, 0x0, 0x0, 0x0)
	/usr/local/go/src/reflect/value.go:321 +0xb4
github.com/clearmatics/autonity/rpc.(*callback).call(0xc0005dc360, 0x14100a0, 0xc000c5e1c0, 0xc000c4f6a0, 0x19, 0x1ce7aa8, 0x0, 0x0, 0x0, 0x0, ...)
	/autonity/build/_workspace/src/github.com/clearmatics/autonity/rpc/service.go:206 +0x2b9
github.com/clearmatics/autonity/rpc.(*handler).runMethod(0xc000cbc090, 0x14100a0, 0xc000c5e1c0, 0xc0001900e0, 0xc0005dc360, 0x1ce7aa8, 0x0, 0x0, 0x0)
	/autonity/build/_workspace/src/github.com/clearmatics/autonity/rpc/handler.go:369 +0x8a
github.com/clearmatics/autonity/rpc.(*handler).handleCall(0xc000cbc090, 0xc000c68ba0, 0xc0001900e0, 0xc0000431d0)
	/autonity/build/_workspace/src/github.com/clearmatics/autonity/rpc/handler.go:331 +0x22c
github.com/clearmatics/autonity/rpc.(*handler).handleCallMsg(0xc000cbc090, 0xc000c68ba0, 0xc0001900e0, 0x13fba01)
	/autonity/build/_workspace/src/github.com/clearmatics/autonity/rpc/handler.go:298 +0x1f3
github.com/clearmatics/autonity/rpc.(*handler).handleMsg.func1(0xc000c68ba0)
	/autonity/build/_workspace/src/github.com/clearmatics/autonity/rpc/handler.go:139 +0x46
github.com/clearmatics/autonity/rpc.(*handler).startCallProc.func1(0xc000cbc090, 0xc000ca4100)
	/autonity/build/_workspace/src/github.com/clearmatics/autonity/rpc/handler.go:226 +0x119
created by github.com/clearmatics/autonity/rpc.(*handler).startCallProc
	/autonity/build/_workspace/src/github.com/clearmatics/autonity/rpc/handler.go:222 +0x66

WARN [09-11|10:51:02.289] Served tendermint_getContractABI         conn=217.138.21.178:57166 reqid=2 t=326.421Âµs err="method handler crashed"
```

**Comment by piersy** [comment]
Hi @flofrie thanks for this info.

Looking at the stack trace at the end of the snippet you posted, I can see that this panic is due to trying to access the blockchain field of the backend object before it has been set.

It used to be the case that the blockchain field was set by the miner when it starts or when the client completes processing the first block it receives from the network, but given that all that is happening concurrently there was really no guarantee of when the blockchain field would be set after starting a node.

This was fixed in this commit 01bdbc04853ee390a518c20ed4bc12ad03d68327, which will be in the next release, which enforces that you cannot have an instance of the backend without having set the blockchain field.

Unfortunately @AdamDynamic did not try to restart the node, If he had I think we would have seen this work eventually. So, I think this is a plausible cause of the issue, but let's leave this ticket open just in case we can collect any further useful information.

@flofrie It would be useful if we (the Autonity team) were able to have access to the ticketing system so that we can jump on these things as soon as they happen, that might allow us to get some better feedback from the person facing the issue.

**Comment by flofrie** [comment]
> @flofrie It would be useful if we (the Autonity team) were able to have access to the ticketing system so that we can jump on these things as soon as they happen, that might allow us to get some better feedback from the person facing the issue.

Yes, in fact you *must* have access to it, so incidents and problems can be escalated to you. I believe at the very least @yazzaoui does have access. Will clarify.

**Comment by yazzaoui** [comment]
Fixed in Autonity 0.6.0 (https://github.com/clearmatics/autonity/pull/596).
